services/121-service/src/payments/fsp-integration/intersolve-visa/dtos/internal/intersolve-visa-card.dto.ts

```diff
@@ -0,0 +1,38 @@
+import { VisaCardAction } from '@121-service/src/payments/fsp-integration/intersolve-visa/enums/intersolve-visa-card-action.enum';
+import { IntersolveVisaCardStatus } from '@121-service/src/payments/fsp-integration/intersolve-visa/enums/intersolve-visa-card-status.enum';
+import { IntersolveVisaTokenStatus } from '@121-service/src/payments/fsp-integration/intersolve-visa/enums/intersolve-visa-token-status.enum';
+import { VisaCard121Status } from '@121-service/src/payments/fsp-integration/intersolve-visa/enums/wallet-status-121.enum';
+import { WrapperType } from '@121-service/src/wrapper.type';
+import { ApiProperty } from '@nestjs/swagger';
+
+export class IntersolveVisaCardDto {
+  @ApiProperty({ example: '123456' })
+  public tokenCode: string;
+
+  @ApiProperty({
+    enum: VisaCard121Status,
+    example: VisaCard121Status.Issued,
+  })
+  public status: WrapperType<VisaCard121Status>;
+
+  @ApiProperty({ example: 'Card issued' })
+  public explanation: string;
+
+  @ApiProperty({ example: '2022-01-01T00:00:00Z' })
+  public issuedDate: Date;
+
+  @ApiProperty()
+  public actions: WrapperType<VisaCardAction[]>;
+
+  @ApiProperty()
+  public debugInformation: WrapperType<IntersolveVisaCardDebugInformation>;
+}
+
+class IntersolveVisaCardDebugInformation {
```

Comment:
Yes, a nitpick, yes thanks for the comment! :)

I prefer the more important code to appear higher in the file. So in this case, I prefer the order as is over what you suggest. It is a pattern we have applied throughout the DTOs and interfaces.

This preference also goes for me for e.g. functions in a class: I prefer public functions higher than private functions. And prefer functions in order of which they would be typically called ("create" before "delete") and/or "bigger" functions higher rather than lower (function for doing a transfer in the fsp service first).

Reply:
Then, I think it's important that this becomes some kind of convention. And there probably needs to be some discussion/voting around defining that convention, for a couple of reasons:

- some of what you're suggesting is _unconventional_ (pun intended), and I would avoid going against those conventions. Some examples:
  - it's a bit of an implicit convention in JS/TS world that private things are declared before public ones
  - it's also implicit convention that "used" things are declared before "the things that use them"
    - this is for historical reasons in JS world because of [hoisting](https://www.w3schools.com/js/js_hoisting.asp#:~:text=Hoisting%20is%20JavaScript's%20default%20behavior%20of%20moving%20all%20declarations%20to,script%20or%20the%20current%20function)) and other nastiness
- There is a bit of a risk in your convention, "I prefer the more important code to appear higher in the file." because the definition of importance is rarely objective

Reply:
I got my preference from Clean Code. Or more precisely: when reading Clean Code it resonated with my experience and opinion.

> In *Clean Code*, Robert C. Martin emphasizes that the order of functions in a class or module should follow a logical sequence that enhances readability and makes the code easier to understand. Specifically, he advocates for an ordering that reads like a narrative or a story. Here's a summary of the key principles related to function order in *Clean Code*:

### 1. **High-Level to Low-Level Order**
   Functions should be organized in a "step-down" approach, where high-level, more abstract functions appear first, followed by lower-level details. This allows the reader to grasp the overall purpose of the code before diving into its implementation specifics.
   
   - **Top-Level Functions:** Should focus on the "what" (i.e., the high-level functionality or business logic).
   - **Lower-Level Functions:** Should focus on the "how" (i.e., details of implementation or helper methods).
   
   By placing higher-level methods at the top, a reader can understand the module's behavior at a glance, without immediately having to dig into the details.

### 2. **Call Order**
   Martin suggests that functions should appear in the order they are called. This means that if function `A` calls function `B`, function `A` should appear before function `B` in the source file. This makes reading the code more intuitive because it flows naturally from the high-level concept (function `A`) down to the supporting details (function `B`).

### 3. **Proximity Principle**
   Functions that are closely related should be placed close to each other. This helps reduce the need to scroll or jump around the file, making the code easier to follow. The idea is to minimize the cognitive load on the reader by keeping related operations together.

### 4. **Private Functions**
   Private or helper functions should be placed near the public functions they support. This proximity reinforces their role as implementation details and keeps them in context with the public interface they assist.

### 5. **Avoid Large Jumps**
   Code should be structured in a way that prevents the reader from having to jump back and forth between distant parts of the file. Functions that belong together logically should be grouped together physically, which also ties into the principle of keeping related functions close.

By following these principles, the codebase becomes more readable and easier to maintain because its structure mirrors how people naturally understand processes—from broad concepts to finer details.

Reply:
Let's have a conversation.

Reply:
Agenda item for chit chat.